lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -1294,6 +1332,14 @@ namespace {
       if (auto *simplified = simplifyTypeConstructionWithLiteralArg(expr))
         return simplified;
 
+      // If we find an unresolved member chain, wrap it in an
+      // UnresolvedMemberChainResultExpr (unless this has already been done).
+      auto *parent = Parent.getAsExpr();
+      if (isMemberChainTail(expr, parent))
+        if (auto *UME = TypeChecker::getUnresolvedMemberChainBase(expr))
+          if (!parent || !isa<UnresolvedMemberChainResultExpr>(parent))
+            return new (ctx) UnresolvedMemberChainResultExpr(expr);
```

Comment:
Why not just save a pointer to the base of the chain in `UnresolvedMemberChainResultExpr` since we have already found it and localize all of the logic in `PreCheckExpression`?

Reply:
Is that compatible in general with the fact that the AST may get rewritten during constraint generation? I was worried about making the `UnresolvedMemberChainResultExpr` depend on the `UnresolvedMemberExpr` nested (potentially) arbitrarily deep in the AST, since if someone introduced logic which swapped out the base for a new expression, the `UnresolvedMemberChainResultExpr` would end up holding on to a stale reference to the old base. Do you think such concerns are overly cautious?

Reply:
Constraint generation shouldn't be modifying AST (expect to a couple of narrow cases like Builtin support) and `UnresolvedMemberExpr` is not going to get replaced with anything regardless, but we could also add an assertion to make sure that this is always true once constraint generator gets to `UnresolvedMemberChainResultExpr`.

Reply:
No, I think if you're not concerned about that possibility it's fine to just store the base on `UnresolvedMemberChainResultExpr`. The assertion would still require us to expose `getMemberChainBase` to both `PreCheckExpression` and `ConstraintGenerator`, so we don't really gain anything by moving it to an assertion.

Reply:
One way to think about it is that toolchain wouldn't actually do the walk since they have assertions disabled, so only debug and assert builds would still maintain the invariant when toolchain wouldn't have the overhead.